test/integration/tcp_proxy_integration_test.cc

```diff
@@ -1218,19 +1227,14 @@ class InjectDynamicMetadata : public Network::ReadFilter {
                 .mutable_filter_metadata())[Envoy::Config::MetadataFilters::get().ENVOY_LB];
       (*map.mutable_fields())[key_] = val;
 
-      // Put this back in the state that TcpProxy expects.
-      read_callbacks_->connection().readDisable(true);
-
       metadata_set_ = true;
     }
     return Network::FilterStatus::Continue;
   }
 
   Network::FilterStatus onNewConnection() override {
-    // TcpProxy disables read; must re-enable so we can read headers.
```

Comment:
Actually we can read-enable here, seems this PR will only buy some time back between the `initialization` and `onNewConnection`?

Reply:
This PR does two things:

1. asks the (TCP) proxy to not read-disable in the first place, so that other filters do not need to know the exact spots where to read-enable and disable again. This enables filters receiving data before and while the upstream connection setup takes place.

2. adds early data buffering to the tcp_proxy filter, so that filters can write (early) data after tcp_proxy has started connect() system call, but the connection has not yet been established. As of now tcp_proxy will assert-fail if it receives data before the connection has been established. This allows filters that read data enabled by point (1) to pass it on without buffering and waiting for connection setup to complete.